src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,164 @@
+
+     /// @notice Constant that defines the share EigenLayer takes of all payments, in basis points
+     uint256 public eigenLayerShareBIPs;
+
+    // TODO: better define this event?
+    event PaymentReceived(address indexed receivedFrom, Payment payment);
+
+    // @notice Emitted when a new Merkle root is posted
+    event NewMerkleRootPosted(MerkleRootPost merkleRootPost);
+
+    event PaymentClaimed(MerkleLeaf merkleLeaf);
+
+    modifier onlyRootPublisher {
+        require(msg.sender == rootPublisher, "PaymentCoordinator: Only rootPublisher");
+        _;
+    }
+
+    function initialize(address _initialOwner, address _rootPublisher, uint256 _eigenlayerShareBips, uint256 _merkleRootActivationDelay)
+        external
+        initializer
+    {
+        _transferOwnership(_initialOwner);
+        _setRootPublisher(_rootPublisher);
+        _setMerkleRootActivationDelay(_merkleRootActivationDelay);
+        _setEigenLayerShareBIPS(_eigenlayerShareBips);
+    }
+
+
+    /**
+     * @notice Makes a payment of sum(amounts) paid in `token`, for `operator`'s contributions to an AVS,
+     * between `startBlockNumber` (inclusive) and `endBlockNumber` (inclusive)
+     * @dev Transfers the total payment from the `msg.sender` to this contract, so the caller must have previously approved
+     * this contract to transfer at least sum(`amounts`) of `token`
+     * @notice Emits a `PaymentReceived` event
+     */
+    function makePayment(Payment calldata payment) external{
+        uint256 sumAmounts;
+        for (uint256 i = 0; i < payment.amounts.length; i++) {
+            sumAmounts += payment.amounts[i];
+        }
+        payment.token.safeTransferFrom(msg.sender, address(this), sumAmounts);
+
+        cumulativeEigenLayerTokeEarnings[payment.token] += sumAmounts * eigenLayerShareBIPs / MAX_BIPS;
+        emit PaymentReceived(msg.sender, payment);
+    }
+
+    // @notice Permissioned function which allows posting a new Merkle root
+    function postMerkleRoot(bytes32 newRoot, uint256 height, uint256 calculatedUpToBlockNumber) external onlyRootPublisher{
+        MerkleRootPost memory newMerkleRoot = MerkleRootPost(newRoot, height, block.number + merkleRootActivationDelay, calculatedUpToBlockNumber);
+        merkleRoots.push(newMerkleRoot);
+        emit NewMerkleRootPosted(newMerkleRoot);
+    }
+
+    /// @notice Permissioned function which allows rootPublisher to nullify a Merkle root
+    function nullifyMerkleRoot(uint256 rootIndex) external onlyRootPublisher{
+        require(block.number <= merkleRoots[rootIndex].confirmedAtBlockNumber, "PaymentCoordinator.nullifyMerkleRoot: Merkle root already confirmed");
+        merkleRoots[rootIndex].root = bytes32(0);
+    }
+
+    // @notice Permissioned function which allows withdrawal of EigenLayer's share of `token` from all received payments
+    function withdrawEigenlayerShare(IERC20 token, address recipient) external onlyRootPublisher{
```

Comment:
I 100% think this should be permissioned to a different address (at least the owner)

Reply:
yeah fair point